tools/apiview/parsers/cpp-api-parser/ApiViewProcessor/ProcessorImpl.cpp

```diff
@@ -0,0 +1,520 @@
+          }
+        }
+      }
+      if (shouldCollect)
+      {
+        // However if the type is in the _internal namespace, then we want to exclude it if we're
+        // excluding internal types.
+        if (typeName.find("::_internal") != std::string::npos)
+        {
+          if (!m_processorImpl->IncludeInternal())
+          {
+            shouldCollect = false;
+          }
+          // If this _internal type is not in a namespace starting with "Azure::Core", then
+          // we want to flag it as an error and include the type.
+          if (typeName.find("Azure::Core") != 0)
+          {
+            m_processorImpl->GetClassesDatabase()->CreateApiViewMessage(
+                ApiViewMessages::InternalTypesInNonCorePackage, typeName);
+            shouldCollect = true;
+          }
+        }
+        // If the type is in the _detail namespace, then we want to exclude it if we're
+        // excluding detail types.
+        if ((typeName.find("::_detail") != std::string::npos) && !m_processorImpl->IncludeDetail())
+        {
+          // There is an exception for Azure::_detail::Clock to the "exclude _detail" rule.
+          if (typeName.find("Azure::_detail::Clock") != 0)
+          {
+            shouldCollect = false;
+          }
+        }
+      }
+      else
+      {
+        // We want to include any free types within the set of headers because they will result in
+        // a diagnostic.
+        PrintingPolicy pp{LangOptions{}};
+        pp.adjustForCPlusPlus();
+        std::string namespaceName;
+        llvm::raw_string_ostream osNamespace(namespaceName);
+        namedDecl->printQualifiedName(osNamespace, pp);
+        std::string typeName;
+        llvm::raw_string_ostream osType(typeName);
+        namedDecl->printName(osType);
+        if (typeName == namespaceName)
+        {
+          shouldCollect = true;
+        }
+      }
+    }
+  }
+  return shouldCollect;
+}
+
+class ApiViewCompilationDatabase : public CompilationDatabase {
+  std::vector<std::filesystem::path> m_filesToCompile;
+  std::filesystem::path m_sourceLocation;
+  std::vector<std::filesystem::path> m_additionalIncludePaths;
+  std::vector<std::string> m_additionalArguments;
+
+public:
+  ApiViewCompilationDatabase(
+      std::vector<std::filesystem::path> const& filesToCompile,
+      std::filesystem::path const& sourceLocation,
+      std::vector<std::filesystem::path> const& additionalIncludePaths,
+      std::vector<std::string> const& additionalArguments)
+      : CompilationDatabase(), m_filesToCompile(filesToCompile),
+        m_sourceLocation(sourceLocation), m_additionalIncludePaths{additionalIncludePaths}
+  {
+    for (auto const& arg : additionalArguments)
+    {
+      m_additionalArguments.push_back(std::string(arg));
+    }
+  }
+  std::vector<std::string>
+      defaultCommandLine{"clang++.exe", "-DAZ_RTTI", "-fcxx-exceptions", "-c", "-std=c++14"};
```

Comment:
Do we make sure this is the full path to exe.  We should prevent running a random exe named clang++.exe.

Reply:
We don't need to - this isn't an actual command line - instead it's a fake command line that we hand the clang tooling. It sees that argv[0] is clang++.exe and starts interpreting all the other command line switches as if they were clang switches.

I'll add a comment.